lib/ast/values.ts

```diff
@@ -1343,21 +1343,37 @@ function dateToSource(date : DateLike) : TokenStream {
 export class DateValue extends Value {
     value : DateLike|null;
 
-    constructor(value : DateLike|null) {
+    constructor(value : DateLike|null, time ?: TimeValue) {
         super(null);
         assert(value === null || isValidDate(value));
+        assert(time === undefined || time === null || (time instanceof TimeValue && time.value instanceof AbsoluteTime));
 
         // a DatePiece with non-null year is actually a fully specified date
         if (value instanceof DatePiece && value.year !== null) {
             let hour = 0, minute = 0, second = 0;
-            if (value.time) {
+            if (time && time.value instanceof AbsoluteTime) {
+                hour = time.value.hour;
+                minute = time.value.minute;
+                second = time.value.second;
+            } else if (value.time) {
                 hour = value.time.hour;
                 minute = value.time.minute;
                 second = value.time.second;
             }
 
             value = new Date(value.year, value.month !== null ? value.month-1 : 0, value.day !== null ? value.day : 1,
                              hour, minute, second);
+        } 
+
+        if (time && time.value instanceof AbsoluteTime) {
+            if (value instanceof Date) {
+                value.setHours(time.value.hour);
+                value.setMinutes(time.value.minute);
+                value.setSeconds(time.value.second);
+            } else if (value instanceof WeekDayDate) {
+                if (value.time) 
+                    value = new WeekDayDate(value.weekday, time.value);
+            }
```

Comment:
I think it's missing the case where value is a DatePiece but the year is null, in which case we need to overwrite the `time` field of DatePiece with the provided time.

It's also missing the case of a DateEdge ("today", "tomorrow", "this week", etc.).
That needs to be treated like a relative time I think - we'll handle that later. But let's add an assert that it won't happen for now.